### PR TITLE
Populate ActionDescriptor EndpointMetadata

### PR DESCRIPTION
**Pull request by martincostello**
**PR Title**

Populate `ActionDescriptor.EndpointMetadata`.

**PR Description**

Populate the `EndpointMetadata` property of the `ActionDescriptor` on `ApiDescription` instances for minimal actions so that the attributes associated with the request delegate can be inspected.

This _seems_ to me like it's something that should be done and is just missing, but in case it was intentional and this use case is intended to be solved in a different way, I've just opened this as a draft for now pending confirmation (or not) of the principle.

Relates to #34061.

### CODE REVIEW

**Comment by martincostello** [comment]
I used the below change to [`SimpleWebSiteWithWebApplicationBuilder`](https://github.com/dotnet/aspnetcore/blob/932c553b5edadc3d4fa233d04ddd1eb839a64772/src/Mvc/test/WebSites/SimpleWebSiteWithWebApplicationBuilder/Program.cs#L11) and the [`JsonResult_Works`](https://github.com/dotnet/aspnetcore/blob/b9efadc59ddbf6743775244622fb95301dd1846c/src/Mvc/test/Mvc.FunctionalTests/SimpleWithWebApplicationBuilderTests.cs#L33-L46) functional test to work through this.

```csharp
app.MapGet("/json", [ApiExplorerSettings(IgnoreApi = true)] (IApiDescriptionGroupCollectionProvider provider) =>
{
    if (provider.ApiDescriptionGroups is not null)
    {
        // Used to inspect and debug ApiDescriptionGroups
    }

    return Json(new Person("John", 42));
});
```

**Comment by martincostello** [comment]
Also, with this and #34063, it's possible there's other things missing from the `ActionDescriptor` that should be being set/populated, but these are the two that I've specifically noticed.

**Comment by martincostello** [comment]
Digging into it more this isn't a magic fix for #34061 for `[ApiExplorerSettings]` to just work, but it seems to be a sensible change as otherwise none of the attributes on the action can be inspected on the `ApiDescription` for customisation by application code.

**Comment by javiercn** [comment]
It's not ok for API explorer to modify the endpoint metadata on action descriptor. The action descriptors should only be modified by the IActionDescriptor providers.

While this potentially solves the problem you are facing, we don't think this is the correct way to solve it.

The endpoint metadata on the action descriptor is "Incomplete" by design and the one on the actual endpoint should be used instead. The metadata on the descriptor only reflects the metadata at the time the descriptor was created and doesn't include anything that has been added after the endpoint was generated. Updating it there, means that other libraries will see a different result after navigating to the swagger page and that is not acceptable.

**Comment by martincostello** [comment]
Then is it the case that part of the implementation in the provider is in the wrong place, given that the `ActionDescriptor` itself is created here?

https://github.com/dotnet/aspnetcore/blob/2032c68402a8cae2bc9e312ddbedc6fa47878b97/src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs#L92-L99

The class modified here implements `IActionDescriptorProvider`, so I don't quite follow the argument that only action descriptor providers should modify it because it is one.

It's also not really modified per se, as it's just completely empty - there is no metadata at all, as it was constructed just a few lines earlier.

The application has no prior access to the `ActionDescriptor` before the method returns because it is the source of it.

If that's still incorrect, then I'll close this PR. In that case though, what is the intended design that would allow the action's metadata to be accessible to the API Explorer for inspection?

**Comment by javiercn** [comment]
@martincostello I might have overlooked something, it's been a while since I've been in that part of the codebase. I'll take another look and update my comment here as necessary.

File: src/Http/Http.Abstractions/src/Routing/EndpointMetadataCollection.cs [hunk 1]
```diff
@@ -151,8 +151,10 @@ public EndpointMetadataCollection(params object[] items)
         /// </summary>
         public struct Enumerator : IEnumerator<object?>
         {
+#pragma warning disable IDE0044
             // Intentionally not readonly to prevent defensive struct copies
-            private readonly object[] _items;
+            private object[] _items;
```
Comment (martincostello) [inline src/Http/Http.Abstractions/src/Routing/EndpointMetadataCollection.cs:156]: @pranavkm This got changed by #34215 but the comment suggests it shouldn't have. I put it back when resolving merge conflicts.

File: src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs [hunk 2]
```diff
@@ -339,6 +341,31 @@ private static void AddResponseContentTypes(IList<ApiResponseFormat> apiResponse
...
 
+        private static void AddActionDescriptorEndpointMetadata(
+            ActionDescriptor actionDescriptor,
+            EndpointMetadataCollection endpointMetadata)
+        {
+            if (endpointMetadata.Count > 0)
+            {
+                // ActionDescriptor.EndpointMetadata is an empty array by
+                // default so need to add the metadata into a new list.
+                var metadata = new List<object>(endpointMetadata.Count);
```
Comment (halter73) [inline src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs:352]: Why not?

```suggestion
                var metadata = new List<object>(endpointMetadata);
```

Is filtering out `null` important? What's adding `null`?

Comment (martincostello) [reply src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs]: Just a habit, have simplified.

**Comment by davidfowl** [comment]
@pranavkm @halter73 What's the deal here?

**Comment by halter73** [comment]
I think we should expose all the method's attributes on the ActionDescriptors generated by EndpointMetadataApiDescriptionProvider somehow, and those are all in RouteEndpoint.Metadata. This makes a lot of sense to me if @pranavkm has no objections.

**Review by halter73 — APPROVED** [review]

**Review by halter73 — APPROVED** [review]

**Review by pranavkm — APPROVED** [review]

**Comment by DamianEdwards** [comment]
I was literally **just** looking for this trying to configure a custom operation id builder for Swashbuckle for minimal APIs. This will let us get to the endpoint name.

**Pull request merged**